Make unit mangling follow paths

keszybz opened the pull request:
See the last commit for overview.

```
tl;dr:                         old                    new
systemctl status /dev           dev.device        dev.mount
systemctl status /../dev       -..-dev.service    dev.mount
systemctl status /../dev/..   -..-dev-...service  -.mount
```

yuwata commented on src/test/test-path-util.c:
```diff
@@ -48,6 +44,74 @@ TEST(path) {
         assert_se(!path_equal_ptr(NULL, "/a"));
 }
 
+TEST(is_path) {
```
Maybe better to add more negative cases? E.g. `"foo"`.

yuwata commented on src/test/test-path-util.c:
```diff
@@ -48,6 +44,74 @@ TEST(path) {
+        assert_se( is_path("/sys/../"));
+}
+
+TEST(is_device_path) {
```
How about `"/usr/../dev/sda"`?

keszybz replied:
It fails, because it does not start with `/dev`. I'll add it to the test case.

yuwata commented on src/basic/path-util.c:
```diff
@@ -1274,7 +1274,8 @@ bool is_device_path(const char *path) {
         /* Returns true for paths that likely refer to a device, either by path in sysfs or to something in
          * /dev. */
```
Please extend the comment that this refuses `/dev`.

keszybz replied:
I extended the comment and added an example.

yuwata commented on src/basic/path-util.c:
```diff
@@ -657,7 +657,14 @@ static int find_executable_impl(const char *name, const char *root, char **ret_f
 }
 
-int find_executable_full(const char *name, const char *root, char **exec_search_path, bool use_path_envvar, char **ret_filename, int *ret_fd) {
+int find_executable_full(
```
Not a matter, but I'd like to split out such not mentioned change into another commit.

keszybz replied:
I split them out to a separate commit. I didn't do that before because these changes I wanted to avoid a trivial commit.

yuwata commented on src/basic/unit-name.c:
```diff
@@ -701,7 +701,13 @@ static bool do_escape_mangle(const char *f, bool allow_globs, char *t) {
  *  If @allow_globs, globs characters are preserved. Otherwise, they are escaped.
  */
-int unit_name_mangle_with_suffix(const char *name, const char *operation, UnitNameMangle flags, const char *suffix, char **ret) {
+int unit_name_mangle_with_suffix(
```
Also this one.

yuwata commented on src/basic/unit-name.c:
```diff
@@ -736,16 +738,23 @@ int unit_name_mangle_with_suffix(
+        if (path_is_absolute(name)) {
+                _cleanup_free_ char *t = NULL;
+
+                r = chase(name, NULL, CHASE_NOFOLLOW|CHASE_NONEXISTENT|CHASE_NO_AUTOFS|CHASE_WARN, &t, NULL);
```
This is a library like function. I'd like to drop `CHASE_WARN`.

keszybz replied:
There's a `flags` argument. Now `CHASE_WARN` is conditionalized on `flags`.

yuwata commented on src/basic/unit-name.c:
```diff
@@ -736,16 +738,23 @@ int unit_name_mangle_with_suffix(
+                if (ERRNO_IS_NEG_RESOURCE(r))
                         return r;
-        }
+                /* If the chase failed, let's just ignore use the unchased path. */
```
Then, maybe better to add `log_debug_errno()` ?

keszybz commented:
Thank you for the review. Updated with changes based on the comments.

yuwata commented on src/basic/path-util.c:
```diff
@@ -1272,9 +1272,16 @@ bool hidden_or_backup_file(const char *filename) {
+         * It does not actually resolve the path.
+         *
+         * Examples:
+         * /dev/sda, /dev/sda/foo, /sys/class, /dev/.., /sys/.., /./dev/foo → yes.
```
`/./dev/foo` is not trivial. Could you add testcase `/./dev/foo` (maybe also `/./dev` and sys variants) ?

yuwata commented on src/basic/unit-name.c:
```diff
@@ -730,22 +732,32 @@ int unit_name_mangle_with_suffix(
+                ChaseFlags chase_flags = CHASE_NOFOLLOW | CHASE_NONEXISTENT | CHASE_NO_AUTOFS |
+                        (warn ? CHASE_WARN : 0);
+
+                r = chase(name, NULL, chase_flags, &t, NULL);
+                log_debug_errno(r, "Failed to chase %s: %m", name);
+                if (ERRNO_IS_NEG_RESOURCE(r))
```
Spurious logging. Triggers assertion if the file does not exsit, as it returns `0`.

keszybz replied:
Arrggh. Should be fixed now.

yuwata reviewed (approved):
LGTM.

poettering commented on src/basic/path-util.c:
```diff
@@ -1272,9 +1272,16 @@ bool hidden_or_backup_file(const char *filename) {
 
-        return PATH_STARTSWITH_SET(path, "/dev/", "/sys/");
+        const char *p = PATH_STARTSWITH_SET(ASSERT_PTR(path), "/dev/", "/sys/");
+        return !isempty(p);
```
I think just doing `return isempty(PATH_STARTSWITH_SET(…))` would be as readable and shorter

poettering commented:
I'd be fine with a logic like this if it never would consult the fs, i.e. if it only would only make redirections we can safely do without actually looking at the fs. i.e. cutting off /../../../ is always safe I guess.

keszybz commented:
OK, updated to just handle leading `/../` (and `/./` in any location, as before). PTAL.

yuwata commented on src/test/test-unit-name.c:
```diff
@@ -234,6 +238,42 @@ TEST(unit_name_mangle) {
+        test_unit_name_mangle_with_suffix_one("/../dev", 1, "dev.mount");
+        test_unit_name_mangle_with_suffix_one("/../dev/.", 1, "dev.mount");
+        /* We don't skip the last '..', and it makes this an invalid device or mount name */
+        test_unit_name_mangle_with_suffix_one("/.././dev/..", 1, "-..-.-dev-...service");
```
Could you add more test cases e.g. `/.././dev`, `/./.././../dev/` ?

yuwata commented on src/basic/unit-name.c:
```diff
@@ -730,22 +730,29 @@ int unit_name_mangle_with_suffix(
+                const char *n = name;
+
+                /* Skip leading /../ */
+                for (const char *m = n; (m = path_startswith(n, "/..")); )
```
Maybe, we should unconditionally do that in path_simplify()?

keszybz commented:
> Maybe, we should unconditionally do that in path_simplify()?

OK, I did that now. This results in quite a bit of churn, so PTAL again.

yuwata commented on src/basic/path-util.h:
```diff
@@ -85,6 +85,19 @@ static inline char* path_simplify(char *path) {
+                        return -ENOMEM;
+        }
+
+        *ASSERT_PTR(ret) = path_simplify(t);
```
In this case I'd like to locate `assert(ret)` at the beginning of the function like the our traditional coding style, to make it easier to understand that the function requires non-NULL `ret`.

yuwata commented on src/basic/path-util.h:
```diff
@@ -85,6 +85,19 @@ static inline char* path_simplify(char *path) {
+static inline int path_simplify_alloc(const char *path, char **ret) {
+        char *t = NULL;
+
+        if (path) {
```
Maybe, better to shortcut when `NULL` path? To make it clear that this function returns NULL when NULL is passed.
```c
assert(ret);

if (!path) {
        *ret = NULL;
        return 0;
}

char *t = strdup(path);
...
```

keszybz replied:
Updated like this.

yuwata reviewed (approved):
Looks nice.
Minor suggestions. Feel free to ignore if you do not like.

YHNdnzj commented:
CentOS CI:

```
14:48:48 testsuite-74.sh[425]: + systemd-escape --path /../hello
14:48:48 testsuite-74.sh[426]: hello
14:48:48 testsuite-74.sh[52]: + echo 'Subtest /usr/lib/systemd/tests/testdata/units/testsuite-74.escape.sh failed'
14:48:48 testsuite-74.sh[52]: Subtest /usr/lib/systemd/tests/testdata/units/testsuite-74.escape.sh failed
```

YHNdnzj commented:
So in `systemd-escape(1)` we document that

> --path
> When escaping or unescaping a string, assume it refers to a file system path. This eliminates leading, trailing or duplicate "/" characters and rejects "."  and ".."  path components.

And after this PR we accept `.` and `..` in `unit_name_path_escape`.

keszybz commented:
> So in `systemd-escape(1)` we document that
> 
> > --path
> > When escaping or unescaping a string, assume it refers to a file system path. This eliminates leading, trailing or duplicate "/" characters and rejects "."  and ".."  path components.
> 
> But after this PR we accept `.` and `..` in `unit_name_path_escape`.

Thanks. I updated the description and tests for path-escape to allow `.` and leading `..`. I think this is reasonable. For `systemd-escape --path` we want to have non-ambiguous paths, but `.` and leading `..` do not create any ambiguity. We already simplified the path, removing duplicate slashes and such, so the argument wasn't *exactly* reversible. So I think the current change, which effectively allows some more non-ambiguous cases, is for the better.

poettering commented on src/basic/path-util.h:
```diff
@@ -85,6 +85,22 @@ static inline char* path_simplify(char *path) {
+                return -ENOMEM;
+
+        *ret = path_simplify(t);
+        return 0;
```
maybe return `1` here?
